File: C#/Collatz.cs
//Author: Jarrett Melnick
//Program to perform a Collatz sequence on a given range and return the top 10 numbers
//with the largest collatz sequences.

using System;

public class Collatz{
	public static void Main(){
		Console.Write("Enter the maximum value to be considered: ");
		long iterate = Convert.ToInt64(Console.ReadLine());

 		number[] result = new number[10]; 

 		for(int i = 0; i < 10; i++){
 			result[i] = new number();
 		}
		//Performs collatz sequence stating from the max number.
		for(long i = 2; i <= iterate; i++){
			long curCollatz = CollatzConjecture(i);
			if((curCollatz > result[9].collatzVal) && !(hasCollatzNumber(result, curCollatz))){
				number num = new number(i, curCollatz);
				result[9] = num;
				result = sorter(result);
			}
		}
		//Prints the top 10 numbers with the largest collatz values.
 		result = sorter(result);
 		for(int i = 0; i < 10; i++){
 			Console.WriteLine(i+1 + " is " + result[i].identity + 
 				" and has a collatz conjecture of " + result[i].collatzVal);
 		}
 	}
 	public static long CollatzConjecture(long val){
 	//Calculates the length of a collatz sequence for a given number
 		long counter = 0;
 		while(val > 1){
 			string binNumber = Convert.ToString(val, 2);
 			val = (binNumber.Trim().EndsWith("1")) ? (3*val+1): (val/2);
 			counter++;
 		}
 		return counter;
 	}
 	private static bool hasCollatzNumber(number[] num, long val){
		//Determines if the given array num has a number with a Collatz sequence of length val.
 		for(int i = 0; i < 10; i++){
 			if(num[i].collatzVal == val)
 			return true;
 		}
 		return false;
 	}
 	private static number[] sorter(number[] num){
 		//Sorts the array of number objects by largest sequence.
 		bool sorted = false;
 		while(!sorted){
 			sorted = true;

 			for(int i = 1; i < 10 ; i++){
 				if(num[i].collatzVal > num[i-1].collatzVal){
 					number temp = num[i-1];
 					num[i-1] = num[i];
 					num[i] = temp;
 					sorted = false;
 				}
 			}
 		}
 		return num;
 	}

 }
 class number{
	//Class to store collatz sequence data.
	public long identity = 0;		//Number's identity
	public long collatzVal = 0;		//Length of Collatz sequence
	public number(){
		identity = 0;
		collatzVal = 0;
	}
	public number(long ident, long colVal){
		//Full constructor
		identity = ident;
		collatzVal = colVal;
	}
	public long getColatzVal(){
		//Returns collatz sequence value
		return collatzVal;
	}
}

File: Julia/Collatz.jl
#Author: Jarrett Melnick
#Program to list the top 10 numbers with the largest Collatz sequences in a given range.
function collatz(val)
    #Function to perform a Collatz sequence on a number and calculate the length of the sequence
    sequence = 1
    while(val > 2)
        if(val % 2 == 1)
            sequence = sequence+1
            val = (3*val+1)
        else
            sequence = sequence + 1
            val = (val >> 1)
        end
    end
    return sequence
end
function smallest(numArray)
    #Returns the index of the smallest value in array numArray
    mini = 1
    for i = 1:10
        if(numArray[i] < numArray[mini])
            mini = i
        end
    end
    return mini
end
function largest(numArray)
    #Returns the index of the largest value in array numArray
    maxi = 1
    for i = 1:10
        if(numArray[i] > numArray[maxi])
            maxi = i
        end
    end
    return maxi
end
function hasSequenceLen(numArray, val)
    #Returns true if the array numArray has the number val
    hasVal = false
    for i = 1:10
        if(numArray[i] == val)
            hasVal = true
        end
    end
    return hasVal
end
function printTable(numArrayI, numArrayS)
    for i = 1:10
        @printf "%d has a length of %d\n" numArrayI[i] numArrayS[i] 
    end
end

#Program logic:
print("Please input the max value to be considered: ")
iterate = readline()
iterate = parse(Int64, chomp(iterate))
identities = [0,0,0,0,0,0,0,0,0,0]
sequences = [0,0,0,0,0,0,0,0,0,0]

println("Calculating Sequences...")
for i = 2:iterate
    curCollatz = collatz(i)
    small = smallest(sequences)
    hasSequence = hasSequenceLen(sequences, curCollatz)
    if((curCollatz > sequences[small]) && !(hasSequence))
        sequences[small] = curCollatz
        identities[small] = i
    end
end
ident2 = copy(identities)
seq2 = copy(sequences)
println("SORTED BY SEQUENCE SIZE")
for i = 1:10
    large = largest(sequences)
    @printf "%d has a Collatz sequence of length %d\n" identities[large] sequences[large]
    identities[large] = 0
    sequences[large] = 0
end
println("SORTED BY INTEGER SIZE")
for i = 1:10
    large = largest(ident2)
    @printf "%d has a Collatz sequence of length %d\n" ident2[large] seq2[large]
    ident2[large] = 0
    seq2[large] = 0
end